oeps/oep-0041-arch-async-server-event-messaging.rst

```diff
@@ -0,0 +1,486 @@
+
+A callback is when you create a message containing an ID or URL that you expect
+the consumer to make a synchronous call to when it receives the message. This is
+commonly used when an event represents some change that is too large to
+practically fit into the 64K message. For instance, we currently emit a generic
+``course_published`` Django signal whenever data is published in Studio, leading
+to a cascade of calls from various apps to the ModuleStore in order to extract
+the content data that they need.
+
+Callbacks threaten performance and stability because they reduce a service's
+ability to control its own load. For instance, a sudden increase in Courseware
+traffic might generate a burst of student analytics events. If this stream of
+events overwhelms your service's ability to consume them, the queue may start to
+back up with unread events. Yet this shouldn't cause your service to fail, since
+it still gets to control how quicky it consumes events off of that queue. It has
+the freedom to either slowly catch up (if the burst was a momentary spike), or
+to scale up additional workers to handle the higher throughput. Your service's
+decision to scale up or down does not directly impact other services.
+
+Things change when we introduce a callback to this same scenario. Say the
+analytics events now include a callback URL to get basic user information. In
+this scenario, doubling the consumers that your service has now also doubles the
+load that your service is placing on the REST endpoint serving this user
+information. There's no way for you as a service owner to know for sure if
+that's safe, and doing so may lead to cascading failures in other systems.
+
+One thing to consider is whether we can emit multiple events that better target
+specific consumer use cases. Let's take the ``course_published`` event as an
+example. Some listeners only care about schedule changes, because they have to
+account for when a course starts and ends. Search indexing really only wants to
+know the exact bit of content that was modified so that it can update that text.
+There is no rule that says a single user action has to translate into a single
+event. Be mindful of what your consumers actually care about and the broad use
+cases you're trying to serve.
+
+If a callback is still necessary, try to make sure that it points to an
+especially robust and performant endpoint. For instance, an event that is fired
+when a user changes their profile information might include a URL to the S3
+location of their new profile picture. Just keep in mind that messages may be
+read long after they're generated, and any presigned S3 URLs you generate might
+be expired by the time a consumer gets them.
+
+Transport Layer
+===============
+
+Events will be sent and received using Kombu, the messaging library that celery
+uses. We will use Redis as the backing server. This allow us to implement this
+messaging scheme without adding any new server dependencies.
```

Comment:
This means that fanout will only be supported with pub/sub: https://github.com/celery/kombu#transport-comparison
In practice, it means that some disconnected clients will miss messages when they reconnect. Isn't that an issue? Mongodb is also part of the Open edX stack -- I think it would be worth considering this alternative.

Reply:
I guess I kind of assumed that we'd use topic exchanges for the most part, as opposed to fanout. I really don't have any operational experience using Redis in this way though, so I could be totally mistaken. Part of the reason why there hasn't been an update to this OEP for a while is because I'm working on a prototype to kick the tires on a few of these things.

Reply:
As for MongoDB... it's beyond the scope of this OEP, but I was hoping that one day we could actually remove MongoDB from the overall stack. There are only two central uses of it right now: Modulestore/Contentstore (where we mostly use it as a key value blob store), and the forums.

That being said, I proposed Redis as the transport mostly because it seemed the easiest thing to administer that's already in our stack, and we already use it as the default for celery. At the code layer, I'd want to just use Kombu and be transport agnostic anyhow. If folks want to use MongoDB or even RabbitMQ, that should be fairly straightforward to set up.

Reply:
It's important to understand that with pub/sub (and thus with Redis) we do not get exactly once delivery, but only at most once delivery. See https://redislabs.com/redis-enterprise/use-cases/messaging/ ("Messaging semantics must be clear")

For instance:

0. subscriber and publisher boot up
1. subscriber crashes
2. publisher emits event
3. event is not received by subscriber
4. subscriber comes back up

This is a very frequent scenario, for instance in deployments without downtime. Sure, you might be doing blue/green deployments properly, but crashes happen nonetheless. If there is a simple syntax error in the subscriber, then all events that will be received will be thrown away. Sure you might guarantee that you have a fallback option that catches errors and re-schedules events in 10 minutes -- but will that work? What if there is an error in the fallback code?

My point is that we will have to engineer a very complex solution just to make up for the failings of the pub/sub model.

I understand that you want to simplify the Open edX, and I can sympathize with that. But please don't displace complexity by moving database features to the code base. 

Apparently, [Redis Streams](https://redis.io/topics/streams-intro) is a redis-backed solution for at least once delivery. I didn't know this alternative to Kafka existed. However, Redis Streams does not seem to be compatible with Kombu. Maybe it's time to reconsider the use of Kombu?

Reply:
I agree that we need to account for subscribers crashing, going away for a while, and then coming back and not missing all the events in the intervening time. I agree that it's a frequent scenario. I have no desire to re-implement a message queue to get things onto our message queue.

My understanding (and this could be wrong), is that Kombu only uses Redis PUB/SUB for fanout exchanges, and that topic exchanges are implemented with Redis lists and sets (i.e. LPUSH, BRPOP), which would then survive client disconnect. I don't know the limits of this because I don't really know the specifics how how Kombu arranges these data structures internally, so I definitely need to test this.

> Apparently, Redis Streams is a redis-backed solution for at least once delivery. I didn't know this alternative to Kafka existed. However, Redis Streams does not seem to be compatible with Kombu. Maybe it's time to reconsider the use of Kombu?

So I am biased here. I helped to write one of the Python drivers for Kafka in the very early days of Datadog, and I'm enormously fond of the project. I think their design trade-offs were incredibly clever, though the ZooKeeper stuff did make me miserable for a few weeks. We've looked at Redis streams briefly when talking about learner analytics tracking logs, as it has many of the features we want without adding something drastically new to our stack. Being able to rewind time and reprocess things is like a crazy super power, and capped streams would be a great compromise between allowing for failure without having data exposed forever.

That being said, I think that event logs like Kafka and Redis Streams are more effort to set up for our common use cases, where we have a wide range of event types and relatively lower throughput. With Kombu, it's really easy to match on a handful of patterns using routing key wildcards, say if you want to audit all GDPR related events throughout the system. Using Kafka or Redis Streams requires more up-front planning about how those streams/topics are going to be arranged.

If it turns out that Redis + Kombu doesn't meet our reliability requirements, I'd probably next see if Kombu + RabbitMQ would work. I'm afraid that adopting Kafka or Redis Streams as a general tool will require us to plan more and write more in terms of code. But I haven't followed the Kafka space closely in a while, so if there are libraries or frameworks that you recommend, I'd be happy to look at them after I do a round of Kombu + Redis testing.

Reply:
> My understanding (and this could be wrong), is that Kombu only uses Redis PUB/SUB for fanout exchanges, and that topic exchanges are implemented with Redis lists and sets (i.e. LPUSH, BRPOP), which would then survive client disconnect. I don't know the limits of this because I don't really know the specifics how how Kombu arranges these data structures internally, so I definitely need to test this.

Yes, this is most likely. The problem with Kombu + Redis (or Rabbitmq for that matter) is that only a single subscriber can receive every message. In other words: there is no fanout. It means that we will not be able to process a single message for different purposes. Let's take an example: when we receive a "subscribed to course" event, we would like both to:

1. warn the course staff about a new subscription
2. forward this event to our analytics system

Without fanout, a worker doing step 1. will have to emit another event in order to be processed by a worker performing step 2.

I am putting Rabbitmq in the same bag as Redis because fanout is implemented by sending multiple copies of the same message to different queues: https://docs.celeryproject.org/projects/kombu/en/stable/reference/kombu.transport.virtual.exchange.html#fanout This really seems backwards once you've played around with Kafka.

It's cool that you've worked with Kafka before! In my experience few people in the python community have. It's a fantastic tool that is really hard to get right, in part because of the Zookeeper requirement, in part because of the [horribly](https://docs.confluent.io/current/installation/configuration/topic-configs.html) [complex](https://docs.confluent.io/current/installation/configuration/producer-configs.html) [configuration](https://docs.confluent.io/current/installation/configuration/consumer-configs.html). But that's a conversation for another time...

> That being said, I think that event logs like Kafka and Redis Streams are more effort to set up for our common use cases, where we have a wide range of event types and relatively lower throughput. With Kombu, it's really easy to match on a handful of patterns using routing key wildcards, say if you want to audit all GDPR related events throughout the system. Using Kafka or Redis Streams requires more up-front planning about how those streams/topics are going to be arranged.

I don't think we need more planning, but a different implementation. With wildcard pattern matching, I expect you wanted to write something like:

```
def wildcard_consumer(event_name, *args, **kwargs):
    event_type = event_name.split(",")[1]
    if event_type == "subscribe":
        subscribe_consumer(*args, **kwargs)
    elif event_type == "unsubscribe":
        unsubscribe_consumer(*args, **kwargs)
    ...
```

Is that right? With Redis Streams, the event type would be passed as an argument of the event:

```
def assigned_consumer(event_name, event_type, *args, **kwargs):
    if event_type == "subscribe":
        subscribe_consumer(*args, **kwargs)
    elif event_type == "unsubscribe":
        unsubscribe_consumer(*args, **kwargs)
    ...
```

So the consumer would only process events from a given stream, but we could pack different events in that stream. Would that work for you?

> If it turns out that Redis + Kombu doesn't meet our reliability/persistence-on-disconnect requirements, I'd probably next see if Kombu + RabbitMQ would work. I'm afraid that adopting Kafka or Redis Streams as a general tool will require us to plan more and write more in terms of code. But I haven't followed the Kafka space closely in a while, so if there are libraries or frameworks that you recommend, I'd be happy to look at them after I do a round of Kombu + Redis testing.

I don't have any experience with Redis Streams, but in my understanding the required tooling is very small. Basically, all we need is Redis v5.0+ and an up-to-date python client. I think it would be a shame to lose the ability to fanout so early in the design process. Also, adding back Rabbitmq to the stack would be a real step backwards, as it was removed more than a year ago.

Reply:
> Yes, this is most likely. The problem with Kombu + Redis (or Rabbitmq for that matter) is that only a single subscriber can receive every message. In other words: there is no fanout. It means that we will not be able to process a single message for different purposes.

I just wrote a couple of experiment scripts, and it does look like we get this capability with Kombu. I set up three instances of [consumers](https://gist.github.com/ormsbee/23616856bbbb11de35dd88f59e984c32) with different queues that connected to the same exchange. One consumer was set to listen for the routing key `org.openedx.catalog.banana.thrown.v1`, another for `*.banana.*`, and the third for `nomatch.key`. When a separate [producer](https://gist.github.com/ormsbee/d5c684b127ffb3accd9cad0d76d070cf) process sent `org.openedx.catalog.banana.thrown.v1` messages, the first two consumers got them while the third did not.

I mean, yes, under the covers it's likely being implemented as pushes to separate queues, but does that really matter for our purposes? The interface is fairly light from the client end. It also seems lighter than having every client examine every message in the stream to filter out the ones they actually care about.

I do have some longer term concerns about robustness with the way Kombu uses redis (I've tested disconnects and catching up once a client reconnects, and things seem to be okay, but I haven't stress tested it). I agree that being able to deliver the same message to multiple consumers is critical, and that we absolutely do not want some crazy chain of clients echoing events out. But based on my reading and experiments, Kombu seems to meet our requirements in these areas.

Would you feel comfortable starting with Kombu if it was kept as a swappable implementation detail? The touch points are fairly minimal (get a message, send a message, ack a message, subscribe to a pattern of messages, maybe some exchange-like concept)–I expect it'll be fairly straightforward to keep those details contained to a small module.

Reply:
@regisb: Just as a heads up: I'm going to remove all mention of the Transport layer from this OEP. Message queue implementation has bubbled up the list of priorities and is actively being worked on this quarter. As part of that, @feanil is leading an investigation/evaluation of transport backends, which will include Redis Streams, Kafka, and Pulsar.

Reply:
Awesome, thanks for the heads-up @ormsbee. I'm curious to see the results of @feanil's research.